### PR TITLE
Enumerate Once Preserving Current Fail Message

### PR DESCRIPTION
**Pull request by jswolf19**
This is to illustrate one strategy to preserve current failed messages while only enumerating an enumerable once. (see PR #759, issue #735)

If this seems to be a viable solution, then tests should probably be added to flush out possible corner cases.

CC @jnm2 @SimonCropp

### CODE REVIEW

**Comment by jswolf19** [comment]
@SimonCropp Is this WIP worth pursuing?

I think I can also override `EnumerableProxy<T>.ToString` to avoid changes in `StringHelper` which may be a more desirable way to go. My only concern is that it would make assumptions about how `StringHelper` works.

Otherwise, I can just add tests for cases when an enumerable is empty.

**Review by jnm2 — CHANGES_REQUESTED** [review]
@jswolf19 I like having all formatting logic in one place in StringHelper the way you have it. That makes it easier to come back to and fully understand without forgetting combo possibilities that are out of sight.

The direction I'd take this in if I did anything would be to make the passing and handling of the buffer beside the original instance even more explicit, not less. But I suspect you struck an efficient balance with this.

File: src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs [hunk 1]
```diff
@@ -57,11 +58,8 @@ public static void ShouldNotBe<T>(this T? actual, T? expected, IEqualityComparer
             bool ignoreOrder,
             string? customMessage)
         {
-            if (actual is not (null or IReadOnlyCollection<T> or ICollection<T> or ICollection))
-                actual = actual.ToList();
-
-            if (expected is not (null or IReadOnlyCollection<T> or ICollection<T> or ICollection))
-                expected = expected.ToList();
+            actual = actual.Wrap();
+            expected = expected.Wrap();
```
[the remark this answers is not in the document]
Comment (jswolf19) [reply src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs:62]: I could get rid of the extension and use `EnumerableProxy<T>.Wrap` here instead, unless you prefer to repeat the checks wherever the Proxy is used.

File: src/Shouldly/Internals/StringHelpers.cs [hunk 2]
```diff
@@ -41,9 +42,17 @@ internal static class StringHelpers
             {
                 var objects = enumerable.Cast<object>();
                 var inspect = "[" + objects.Select(o => o.ToStringAwesomely()).CommaDelimited() + "]";
-                if (inspect == "[]" && value.ToString() != objectType.FullName)
+                if (inspect == "[]")
                 {
-                    inspect += " (" + value + ")";
+                    if (value is IProxy proxy)
```
[the remark this answers is not in the document]
Comment (jswolf19) [reply src/Shouldly/Internals/StringHelpers.cs:47]: The reason I went with `IProxy` is because, at creation and test time, the type `IEnumerable<T>` is necessary, but `StringHelper` doesn't have or need the type argument, so a non-generic type is ideal. Since this hack doesn't need to be confined to `IEnumerable`, I made the interface a more generic name. I was also wondering thinking of moving it outside of the enumerable-specific logic, but did this for now to get something quick for discussion. If the ability to proxy values has a broader use-case, then I wouldn't mind making it a more generalized approach.

[the remark this answers is not in the document]
Comment (jnm2) [reply src/Shouldly/Internals/StringHelpers.cs:47]: Good point, I missed that `IProxy` is currently essential. Its naming convention (it's not `IEnumerableProxy`) makes it look like something I would expect to be able to use for non-enumerable situations, so that is what makes me want this outside the `TryGetEnumerable` check.

File: src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jnm2) [reply src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs:62]: I'm sorry, I misunderstood the first few times I came to this code and thought that the `Wrap` method always wrapped. If you call it `WrapIfNonCollection` then I don't mind the checks moving into that method.

[the remark this answers is not in the document]
Comment (jswolf19) [reply src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs:62]: Actually, collections are wrapped as well, I just don't call `ToList` on them. The only time no wrapping occurs is if the value is null or if it is already wrapped. The idea was that you could call `Wrap` at all entry points, without having to worry about excessive wrapping if calls are chained.

[the remark this answers is not in the document]
Comment (jnm2) [reply src/Shouldly/ShouldlyExtensionMethods/ShouldBe/GenericShouldBeTestExtensions.cs:62]: My personal preference if anything would be to throw an exception if Wrap was called on an already-wrapped collection, because if you find yourself wrapping a wrapper, the most likely scenario is that you're making a larger programming mistake.

I'm also not sure I like the concept of wrapping collections (as opposed to only wrapping non-collections), but I feel less strongly there.

**Review by jnm2 — APPROVED** [review]
Thanks, this looks great!

**Comment by jnm2** [comment]
Well thank you, we appreciate the assistance!
